Main/HW26/AIMoveBestRouteComponent.cpp

```diff
@@ -0,0 +1,159 @@
+
+AIMoveBestRouteComponent::AIMoveBestRouteComponent(GameObject* referenceObject, GameObject* targetObject)
+    : AIMoveToComponent(referenceObject, targetObject), m_mapGraph(buildGraph())
+{
+#ifdef DEBUG_VERTEX
+    m_text.setFont(FontManager::getInstance().getDefaultFont());
+
+    m_text.setCharacterSize(48); // in pixels, not points!
+    m_text.setFillColor(sf::Color::Blue);
+    m_text.setOrigin(m_text.getGlobalBounds().getSize().x / 2, m_text.getGlobalBounds().getSize().y / 2);
+#endif
+}
+
+void AIMoveBestRouteComponent::update(float dt)
+{
+    updatePath(m_mapGraph);
+    const sf::Vector2u nextTileToMove = getNextTileToMove();
+    const sf::Vector2f targetPosition = Map::GetInstance().getPositionFromTile(nextTileToMove);
+
+    moveTo(dt, targetPosition);
+}
+
+void AIMoveBestRouteComponent::draw(sf::RenderWindow* window)
+{
+#ifdef DEBUG_VERTEX
+    const Map::TilesMapType& mapTiles = Map::GetInstance().getAllTiles();
+
+    for (unsigned y = 0; y < mapTiles.size(); y++)
+    {
+        for (unsigned x = 0; x < mapTiles[y].size(); x++)
+        {
+            const int vertixNumber = convertMapTileToVertix({ x, y });
+            const sf::Vector2f position = Map::GetInstance().getPositionFromTile({ x, y });
+
+            m_text.setPosition(position);
+            m_text.setString(std::to_string(vertixNumber));
+
+            window->draw(m_text);
+        }
+    }
+#endif 
+
+#ifdef DEBUG_PATH
+    for (auto vertex : m_path)
+    {
+        sf::RectangleShape path = sf::RectangleShape({ 30, 30 });
+
+        path.setPosition(Map::GetInstance().getPositionFromTile(convertVertixToMapTile(vertex)));
+        path.setFillColor(sf::Color::Yellow);
+
+        window->draw(path);
+    }
+#endif
+}
+
+//#TODO, STUDENTS: Build graph from map Tiles
+Graph AIMoveBestRouteComponent::buildGraph()
+{
+    const Map::TilesMapType& mapTiles = Map::GetInstance().getAllTiles();
+    const unsigned numRows = mapTiles.size();
+    const unsigned numCols = mapTiles[0].size();
+    Graph graph(numRows * numCols);
+
+    //Traverse all map tiles
+    //Check if it's passable
+    //IF it's - check if its neighbours up/down/left/right are passable
+    //if so - add edges between those neighbours to the graph
+    // using convertMapTileToVertix function
+
+    const std::vector<std::pair<int, int>> directions = {
+            { 0, -1 }, // Up
+            { 0, 1 },  // Down
+            { -1, 0 }, // Left
+            { 1, 0 }   // Right
+    };
+
+    for (unsigned y = 0; y < numRows; ++y) {
+        for (unsigned x = 0; x < numCols; ++x) {
+            if (!mapTiles[y][x]) { // Tile is passable
+                const unsigned currentVertex = convertMapTileToVertix({x, y});
+
+                for (const auto& [dx, dy] : directions) {
+                    const int newX = x + dx;
+                    const int newY = y + dy;
+
+                    if (newX >= 0 && newX < numCols && newY >= 0 && newY < numRows && !mapTiles[newY][newX]) {
```

Comment:
I would split these conditions into 2 just for the sake of readability and splitting concerns:
validation and map check